### ISSUE TITLE
The sqrt drop-down did not work in Chrome

### ISSUE THREAD

**Issue by gherson** [jmoenig/Snap--Build-Your-Own-Blocks#753]
The sqrt operator's drop-down is not working for me in Google Chrome 41.0: the drop-down doesn't drop or change when clicked.  I then refreshed the page, no change.  

The same drop-down does work for me in Internet Explorer 11.0.  And it now does in Chrome when I opened a new window.  Does that make this a non-issue?  

Operating System: Windows 7 Pro.

**Comment by cycomachead** [comment]
What type of device were you using to click? (Mouse, trackpad, touch screen)?

I believe this is related to clicks being interpreted as drags. I think it might be time to implement a small drag threshold.

**Comment by jmoenig** [comment]
It doesn't sound like a problem that could be solved by introducing a dragging threshold. The point is that the click event *does* work in a new tab, so it must be a bug, an exception even. Something must happen to Morphic here that prevents the click. Actually, the most likely cause would be a (JS) error.

@gherson if that happens again, could you open a console in the Chrome tab and let us know whether there is any error message there? And if there isn't it might be because Snap is catching it. So, if you still can - could you then switch to dev mode by shift-clicking on the Snap logo and selecting the red "Switch to dev mode" menu entry - then again try clicking on the drop-down and then check the JS console again? That would probably be helpful. Thanks!

**Comment by cycomachead** [comment]
> It doesn't sound like a problem that could be solved by introducing a dragging threshold. The point is that the click event does work in a new tab, so it must be a bug, an exception even. Something must happen to Morphic here that prevents the click.

It could also be a coincidence that refereshing a tab and trying to click again led to a successful click register. At least once or twice (with non-mac devices) I've hard that happen - where it's just kind of random.

**Comment by brianharvey** [comment]
I vote for a drag threshold whether or not it solves this particular problem.  I'm forever trying to change the name of a variable and getting a copy instead.  I'd make it a *large* threshold -- you should have to drag the copy all the way off the original for it to stick around.

**Comment by jmoenig** [comment]
sorry guys, but if anything I want to make Snap! *more* responsive to user input, not less. If *drag* and *click* are too close it's better to implement to notion of a *focus* or to find a different gesture altogether than to prefer one over the other. Also, I suspect that often people are setting their screen resolutions *way* higher than is good for them. Since blocks are *interactive* widgets you cannot compare them to text, so we might want to thing about perhaps making them significantly bigger all the time.

**Comment by jmoenig** [comment]
Oh, and btw, if your mouse is too sensitive you can change that in your system's settings!!!

**Comment by cycomachead** [comment]
> I'd make it a large threshold -- you should have to drag the copy all the way off the original for it to stick around.

I think that UI issue probably deserves a separate threshold than a default one, but I whole heartedly agree here. I've often created copies of variables but am not able to see that I've done it. 

Besides, in general, there are plenty of OS functions which have drag thresholds. 

> Oh, and btw, if your mouse is too sensitive you can change that in your system's settings!!!

I don't think this is a valid option for a few reasons -- schools often don't allow students to change settings, and the issue is ususally an issue of precision of clicking, which is something users usually can't control, especially if the pointer is in fact their own finger.
The fact that 90% of the reports are from Windows users definitely says something about either Windows or other hardware.

**Comment by gherson** [comment]
thanks for the attention, all!
I will definitely attempt the error catching steps @jmoenig outlined next time i have a similar problem and report back.  no problem currently.
I use a trackpad.

**Comment by gherson** [comment]
I just had the same issue.  The solution is to be patiently tap lightly on my touchpad.

**Comment by gherson** [comment]
Patience exceeded so I opened a javascript console in Chrome.  There are no errors.  Switched to Dev Mode, still no errors.  (I just see "Synchronous XMLHttpRequest on the main thread is deprecated because of its detrimental effects to the end user's experience.")

**Comment by jmoenig** [comment]
Which platform do you run Chrome on? Mac, Windows, Linux?

Also, when your Snap is in a state where the drop-down doesn't work, is there anything else that's strange? For example: Instead of popping up a menu does the script get fired? Other menus still work, right, because you were able to switch to dev mode? Is it just this instance of the sqrt block that exhibits the problem, or do drop-downs in general stop working? Are you still able to drag the sqrt and exchange it for another one that works?

Thanks!

**Comment by jmoenig** [comment]
Oh, here's another idea: Do you have any Chrome toolbar utilities installed? Especially the Aperture toolbar? One thing that has happened occasionally in the past is that some utilities modify the Snap page's html to insert other DOM elements on the page, and *this* slightly offsets the mouse-coordinates for the Snap canvas. So, when that happens again, could you perhaps take a screenshot and post it here? Also, can you check whether the given mouse coordinates look "right"? (Can you still switch to the "sensing" category, check the watchers for the mouse coordinates and do they make sense (is 0, 0 still the center of the stage)?

**Comment by gherson** [comment]
The best I can describe the problem (other than creating a laundry list) is that the small fiddly items on Snap! become 90-99% unresponsive.  E.g., the sqrt drop-down is only giving me a flash of shadow when clicked.  (Finished "Tic Tac Toe" using IE. (Shared under username gherson.))

0, 0 is still center stage.  Operating System: Windows 7 Pro.  I just disabled all my Chrome extensions.  Aperture was not one of them.

Screennshot:
![tictactoe](https://cloud.githubusercontent.com/assets/1238786/6917195/6942d636-d770-11e4-92d4-89b1303e85eb.png)

**Comment by jmoenig** [comment]
Thanks! So, what about the bigger items in Snap? Do they also become less responsive than they used to? Can you still drag and drop sprites, other blocks, and does mousing over the buttons in the toolbar and over the category tabs still feel responsive, or is that slowing down, too? What about scrolling, does that still work? So, if I understand it right it's not just a *single* sqrt block that becomes less responsive, but *every* little item, including e.g. the expansion arrows for a "list" block, or for rings, "script variables" etc. Is That right?

It doesn't appear to be a hard JS bug, or an exception. So, maybe it could be a scheduling issue. I notice that you're using Windows, and most reports we've been getting for this problem occurred on Windows machines.

Roping @bromagosa, @cycomachead and @awangenh into this discussion I could imagine that this issue of Snap becoming "unresponsive" after a while might be related to my currently over-eager clock scheduling. The browser might simply "choke", and the remedy might be - as apparently works well on less powerful hardware - to gently slow the clock down, which might not even slow down Snap itself.

Bottom line, I'll try to come up with a patch that will let us experiment with various clock-speeds, and this might just be the real reason of what's going wrong here....

**Comment by gherson** [comment]
left my little game running for hours but nothing has changed (tab wasn't refreshed, etc) and the problem is not currently occurring.  I'll wait for it to reappear, then reply to your questions. Thank you.

**Comment by gherson** [comment]
Answers to each question:

"Can you still drag and drop sprites"?    Yes
"other blocks"?    Yes
"does mousing over the buttons in the toolbar 
and over the category tabs still feel responsive, 
or is that slowing down, too?"     Responsive. Switching palettes can take a half second.
"What about scrolling, does that still work?"    Works fine.

"So, if I understand it right it's not just a single sqrt block that becomes less responsive, but every little item, including e.g. the expansion arrows for a "list" block, or for rings, "script variables" etc."
Every block's drop-down is currently working very poorly.  And I don't see anything else that's a problem.

**Comment by jmoenig** [comment]
Thank you for this detailed answer, @gherson !

To sum it up, this issue appears to be pretty much isolated to drop-downs in block input slots but only occur after a while. I'll investigate!

**Comment by gherson** [comment]
appreciated.

**Comment by jmoenig** [comment]
I'm keeping this open, because I want to actually fix it :-)

**Comment by gherson** [comment]
my bad, sorry.  i hope you can figure out a way to reproduce.

**Comment by stirlingstout** [comment]
Just to add something to this: I get exactly the same problem on Chrome (Version 41.0.2272.118 m) but not on IE (Version: 11.0.9600.17690) but the images inside IE is quite a lot bigger. It's not restricted to the sqrt dropdown but any dropdown.

Opening a new window from http://snap.berkeley.edu (not a new tab) seems to fix the problem.

The horizontal pop-outs (e.g., on [list] and [run]) work fine.

**Comment by wortzmanb** [comment]
I've seen this issue a number of times, and while I haven't investigated deeply, I've noticed a few things:
  - It mostly, but not exclusively, occurs in Chrome.
  - There typically is or has been another tab open, often with a document of some kind viewed in it (like a PDF).
  - Clearing the browser cache and restarting the browser often fixes the problem, albeit temporarily.

If there's anything I can do to help here, please let me know.  This is becoming an annoyance in my classroom, where I often have students with Snap! open in one tab and the instructions for the day's lab open in another.  

Thanks!

**Comment by cycomachead** [comment]
This should have been fixed with 4.0.3 - if it's not we can reopen this and discuss again!
